### PR TITLE
W-11364621: Add DW validation to AST validation for flow-refs (lookup) in a DW script

### PR DESCRIPTION
**Pull request by mbuchwald**

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/internal/el/ExpressionLanguageAdaptor.java [hunk 1]
```diff
@@ -156,4 +158,9 @@ Iterator<TypedValue<?>> split(String expression, CoreEvent event, BindingContext
    * @since 4.3.0
    */
   CompiledExpression compile(String expression, BindingContext bindingContext);
+
+  default ScopePhaseValidationMessages collectScopePhaseValidationMessages(String script, String nameIdentifier,
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/el/ExpressionLanguageAdaptor.java:162]: add javadoc

File: runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/OperationValidation.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+import org.mule.runtime.api.component.ComponentIdentifier;
+import org.mule.runtime.ast.api.ComponentAst;
+import org.mule.runtime.ast.api.validation.Validation;
+
+import java.util.List;
+import java.util.function.Predicate;
+
+
+public abstract class OperationValidation implements Validation {
```
Comment (elrodro83) [inline runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/OperationValidation.java:21]: rename to `MuleSdkOperationValidation`

File: runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/OperationDoesNotHaveForbiddenComponent.java [hunk 3]
```diff
@@ -19,16 +15,11 @@
...
 import org.mule.runtime.ast.api.validation.ValidationResultItem;
 
-import java.util.List;
 import java.util.Optional;
-import java.util.function.Predicate;
 
-public abstract class OperationDoesNotHaveForbiddenComponent implements Validation {
-
-  private static final ComponentIdentifier OPERATION_IDENTIFIER = builder().namespace("operation").name("def").build();
+public abstract class OperationDoesNotHaveForbiddenComponent extends OperationValidation {
```
Comment (elrodro83) [inline runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/OperationDoesNotHaveForbiddenComponent.java:22]: rename to `MuleSdkOperationDoesNotHaveForbiddenComponents`

File: runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/OperationDoesNotHaveDeprecatedExpression.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+import org.mule.runtime.ast.api.ComponentAst;
+import org.mule.runtime.ast.api.validation.ValidationResultItem;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Stream;
+
+public class OperationDoesNotHaveDeprecatedExpression extends OperationValidation {
```
Comment (elrodro83) [inline runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/OperationDoesNotHaveDeprecatedExpression.java:33]: rename to `MuleSdkOperationDoesNotHaveForbiddenFunctionsInExpressions`

File: runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/OperationDoesNotHaveDeprecatedExpression.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  private static final MuleVersion DEPRECATE_UP_TO = new MuleVersion("2.5.0");
+  private static final Set<String> DEPRECATION_DOMAINS = ImmutableSet.of("dw", "Mule");
+
+  public OperationDoesNotHaveDeprecatedExpression(ExpressionLanguage el) {
+    this.expressionLanguage = el;
+  }
+
+  @Override
+  public String getName() {
+    return "Operation doesn't have deprecated expressions";
```
Comment (elrodro83) [inline runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/OperationDoesNotHaveDeprecatedExpression.java:48]: reword to disambiguate `Operation`

File: runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/OperationDoesNotHaveDeprecatedExpression.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  @Override
+  public String getDescription() {
+    return "The operation does not have a function as deprecated prior to DW 2.5 within itself";
```
Comment (elrodro83) [inline runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/OperationDoesNotHaveDeprecatedExpression.java:53]: reword to disambiguate `Operation`

File: runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/OperationDoesNotHaveForbiddenComponent.java [hunk 7]
```diff
@@ -42,16 +33,6 @@ public String getDescription() {
     return format("Operations cannot have a '%s' component within", forbiddenComponentIdentifier());
```
Comment (elrodro83) [inline runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/OperationDoesNotHaveForbiddenComponent.java:33]: reword to disambiguate `Operation`

File: runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/OperationDoesNotHaveDeprecatedExpression.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        && isDeprecableProduct(warning));
+  }
+
+  /**
+   * @return All the (raw) expressions within this {@link ParameterizedModel}
+   */
+  private Stream<String> getAllExpressions(ParameterizedModel model, ComponentAst componentAst) {
+    return model.getParameterGroupModels().stream()
+        .flatMap(groupModel -> groupModel.getParameterModels().stream()
+            .filter(parameter -> !parameter.getExpressionSupport().equals(ExpressionSupport.NOT_SUPPORTED))
```
Comment (elrodro83) [inline runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/OperationDoesNotHaveDeprecatedExpression.java:84]: static import

File: runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/OperationDoesNotHaveDeprecatedExpression.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+            .map(parameter -> componentAst.getParameter(groupModel.getName(), parameter.getName()).getRawValue())
+            .filter(Objects::nonNull));
+  }
+
+  @Override
+  public Optional<ValidationResultItem> validate(ComponentAst component, ArtifactAst artifact) {
+    if (component.recursiveStream().anyMatch(componentAst -> componentAst.getModel(ParameterizedModel.class)
+        .map(model -> getAllExpressions(model, componentAst).anyMatch(this::isInvalidExpression)).orElse(false))) {
+      return of(create(component, this,
+                       "Using an invalid function within an operation. All functions deprecated up to DataWeave 2.5 cannot be used inside a Mule Operation"));
```
Comment (elrodro83) [inline runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/OperationDoesNotHaveDeprecatedExpression.java:94]: reword to disambiguate `Mule Operation`

File: runtime-extension-model/src/test/java/org/mule/runtime/config/internal/validation/OperationDoesNotHaveDeprecatedExpressionTestCase.java [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+
+@Features({@Feature(MULE_DSL), @Feature(REUSE)})
+@Stories({@Story(DSL_VALIDATION_STORY), @Story(OPERATIONS)})
+public class OperationDoesNotHaveDeprecatedExpressionTestCase extends AbstractCoreValidationTestCase {
```
Comment (elrodro83) [inline runtime-extension-model/src/test/java/org/mule/runtime/config/internal/validation/OperationDoesNotHaveDeprecatedExpressionTestCase.java:49]: add a test case where a deprecated expression is not used within a mule sdk operation as well

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/MuleSdkOperationDoesNotHaveForbiddenFunctionsInExpressions.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+            .map(parameter -> componentAst.getParameter(groupModel.getName(), parameter.getName()).getRawValue())
+            .filter(Objects::nonNull));
+  }
+
+  @Override
+  public Optional<ValidationResultItem> validate(ComponentAst component, ArtifactAst artifact) {
+    if (component.recursiveStream().anyMatch(componentAst -> componentAst.getModel(ParameterizedModel.class)
+        .map(model -> getAllExpressions(model, componentAst).anyMatch(this::isInvalidExpression)).orElse(false))) {
+      return of(create(component, this,
+                       "Using an invalid function within a Mule SDK operation. All functions deprecated up to DataWeave 2.5 cannot be used inside a Mule Operation"));
```
Comment (elrodro83) [inline runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/MuleSdkOperationDoesNotHaveForbiddenFunctionsInExpressions.java:94]: add the location of the invalid call within the script, refer to `AbstractExpressionSyntacticallyValid `

File: runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/MuleSdkOperationDoesNotHaveForbiddenFunctionsInExpressions.java [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+        && isDeprecableProduct(warning));
+  }
+
+  /**
+   * @return All the (raw) expressions within this {@link ParameterizedModel}
+   */
+  private Stream<String> getAllExpressions(ParameterizedModel model, ComponentAst componentAst) {
+    return model.getParameterGroupModels().stream()
+        .flatMap(groupModel -> groupModel.getParameterModels().stream()
+            .filter(parameter -> !parameter.getExpressionSupport().equals(ExpressionSupport.NOT_SUPPORTED))
```
Comment (elrodro83) [inline runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/MuleSdkOperationDoesNotHaveForbiddenFunctionsInExpressions.java:84]: static import

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/mule/runtime/core/internal/el/ExpressionLanguageAdaptor.java [hunk 13]
```diff
@@ -156,4 +158,16 @@ Iterator<TypedValue<?>> split(String expression, CoreEvent event, BindingContext
    * @since 4.3.0
    */
   CompiledExpression compile(String expression, BindingContext bindingContext);
+
+  /**
+   * @param script the script to be validated.
+   * @param nameIdentifier the name of the script file.
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/el/ExpressionLanguageAdaptor.java:164]: what if tghe script is inline?

File: runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/MuleSdkOperationDoesNotHaveForbiddenFunctionsInExpressions.java [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+  private List<ComponentParameterAst> getAllExpressions(ParameterizedModel model, ComponentAst componentAst) {
+    return model.getParameterGroupModels().stream()
+        .flatMap(groupModel -> groupModel.getParameterModels().stream()
+            .filter(parameter -> !parameter.getExpressionSupport().equals(NOT_SUPPORTED))
+            .map(parameter -> componentAst.getParameter(groupModel.getName(), parameter.getName())))
+        .filter(param -> param.getRawValue() != null && isExpression(param.getRawValue()))
+        .collect(toList());
+  }
+
+  private Map<String, String> locationToAdditionalData(Location location) {
```
Comment (elrodro83) [inline runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/MuleSdkOperationDoesNotHaveForbiddenFunctionsInExpressions.java:94]: this also exists in the syntactic validation, can it be extracted?

File: runtime-extension-model/src/test/java/org/mule/runtime/config/internal/validation/MuleSDKOperationDoesNotHaveDeprecatedExpressionTestCase.java [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+    final Optional<ValidationResultItem> msg = runValidation(XML_NAMESPACE_DEF +
+        "<operation:def name=\"someOp\"><operation:body>" +
+        "<set-payload value=\"10\"/>" +
+        "<logger level=\"WARN\" message=\"#[payload ++ 10]\"/>" +
+        "<set-payload value=\"#[payload + 1]\"/>" +
+        "<logger level=\"WARN\" message=\"#[%dw 2.0 --- { 'v' : Mule::lookup('someFlow') }]\"/>" +
+        "<set-variable value=\"#[someFunction(10)]\"/>" +
+        "</operation:body></operation:def>" +
+        XML_CLOSE).stream().findFirst();
+    assertThat(msg, is(CoreMatchers.not(empty())));
```
Comment (elrodro83) [inline runtime-extension-model/src/test/java/org/mule/runtime/config/internal/validation/MuleSDKOperationDoesNotHaveDeprecatedExpressionTestCase.java:158]: static import

File: runtime-extension-model/src/test/java/org/mule/runtime/config/internal/validation/MuleSDKOperationDoesNotHaveDeprecatedExpressionTestCase.java [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+    when(expressionLanguage.collectScopePhaseValidationMessages(contains("Mule::lookup("), anyString(), any()))
+        .thenReturn(new TestScopePhaseValidationMessages(singletonList(new TestScopePhaseValidationItem("2.5.0",
+                                                                                                        "Mule::lookup"))));
+    final Optional<ValidationResultItem> msg = runValidation(XML_NAMESPACE_DEF +
+        "<operation:def name=\"someOp\"><operation:body>" +
+        "<set-payload value=\"10\"/>" +
+        "<logger level=\"WARN\" message=\"#[%dw 2.0 --- { 'v' : Mule::lookup('someFlow') }]\"/>" +
+        "</operation:body></operation:def>" +
+        XML_CLOSE).stream().findFirst();
+    assertThat(msg, is(CoreMatchers.not(empty())));
```
Comment (elrodro83) [inline runtime-extension-model/src/test/java/org/mule/runtime/config/internal/validation/MuleSDKOperationDoesNotHaveDeprecatedExpressionTestCase.java:107]: static import

File: runtime-extension-model/src/test/java/org/mule/runtime/config/internal/validation/MuleSDKOperationDoesNotHaveDeprecatedExpressionTestCase.java [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+        .thenReturn(new TestScopePhaseValidationMessages(singletonList(new TestScopePhaseValidationItem("2.3.0",
+                                                                                                        "dw::core::Objects::keySet"))));
+    final Optional<ValidationResultItem> msg = runValidation(XML_NAMESPACE_DEF +
+        "<operation:def name=\"someOp\"><operation:body>" +
+        "<set-payload value=\"10\"/>" +
+        "<logger level=\"WARN\" message=\"#[%dw 2.0 import * from dw::core::Objects --- { 'keySet' : keySet({ 'a' : true, 'b' : 1}) }]\"/>"
+        +
+        "</operation:body></operation:def>" +
+        XML_CLOSE).stream().findFirst();
+    assertThat(msg, is(CoreMatchers.not(empty())));
```
Comment (elrodro83) [inline runtime-extension-model/src/test/java/org/mule/runtime/config/internal/validation/MuleSDKOperationDoesNotHaveDeprecatedExpressionTestCase.java:92]: static import

File: runtime-extension-model/src/test/java/org/mule/runtime/config/internal/validation/MuleSDKOperationDoesNotHaveDeprecatedExpressionTestCase.java [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+
+@Features({@Feature(MULE_DSL), @Feature(REUSE)})
+@Stories({@Story(DSL_VALIDATION_STORY), @Story(OPERATIONS)})
+public class MuleSDKOperationDoesNotHaveDeprecatedExpressionTestCase extends AbstractCoreValidationTestCase {
```
Comment (elrodro83) [inline runtime-extension-model/src/test/java/org/mule/runtime/config/internal/validation/MuleSDKOperationDoesNotHaveDeprecatedExpressionTestCase.java:46]: add a test case asserting the location coordinates of the failure

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: runtime-extension-model/src/test/java/org/mule/runtime/config/internal/validation/MuleSDKOperationDoesNotHaveDeprecatedExpressionTestCase.java (hunk 15 above)
Comment (mbuchwald) [reply runtime-extension-model/src/test/java/org/mule/runtime/config/internal/validation/MuleSDKOperationDoesNotHaveDeprecatedExpressionTestCase.java:158]: I am already using another "not" (AdditionalMatchers)

File: runtime-extension-model/src/test/java/org/mule/runtime/config/internal/validation/MuleSDKOperationDoesNotHaveDeprecatedExpressionTestCase.java (hunk 16 above)
Comment (mbuchwald) [reply runtime-extension-model/src/test/java/org/mule/runtime/config/internal/validation/MuleSDKOperationDoesNotHaveDeprecatedExpressionTestCase.java:107]: Same as previous

File: runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/MuleSdkOperationDoesNotHaveForbiddenFunctionsInExpressions.java [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+  }
+
+  @Override
+  public List<ValidationResultItem> validateMany(ComponentAst component, ArtifactAst artifact) {
+    return getExpressionsWithWarnings(component).stream()
+        .map(pair -> create(component, pair.getFirst(), this,
+                            "Using an invalid function within a Mule SDK operation. All functions deprecated up to DataWeave 2.5 cannot be used inside a Mule Operation. Expression: "
+                                + pair.getFirst().getRawValue(),
+                            locationToAdditionalData(pair.getSecond().getLocation())))
+        .collect(Collectors.toList());
```
Comment (elrodro83) [inline runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/MuleSdkOperationDoesNotHaveForbiddenFunctionsInExpressions.java:110]: static import

File: runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/ValidationUtils.java [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.config.internal.validation;
+
+import org.mule.runtime.api.el.validation.Location;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.mule.runtime.config.api.validation.ExpressionsSyntacticallyValidAdditionalDataKeys.*;
```
Comment (elrodro83) [inline runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/ValidationUtils.java:14]: no `*`

File: runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/ValidationUtils.java [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+
+public class ValidationUtils {
```
Comment (elrodro83) [inline runtime-extension-model/src/main/java/org/mule/runtime/config/internal/validation/ValidationUtils.java:16]: package visibility

File: core/src/main/java/org/mule/runtime/core/internal/el/ExpressionLanguageAdaptor.java [hunk 22]
```diff
@@ -156,4 +158,17 @@ Iterator<TypedValue<?>> split(String expression, CoreEvent event, BindingContext
    * @since 4.3.0
    */
   CompiledExpression compile(String expression, BindingContext bindingContext);
+
+  /**
+   * @param script         the script to be validated.
+   * @param nameIdentifier the name of the script file. If the script is inline, it can be an empty string.
+   * @param bindings       the {@link TypeBindings} to be considered for the validation.
+   * @return A {@link ScopePhaseValidationMessages} with the result of the validations perform to the script, considering
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/el/ExpressionLanguageAdaptor.java:166]: `perform to the script` -> `performed on the script`

**Review by elrodro83 — CHANGES_REQUESTED** [review]

**Pull request merged**